### PR TITLE
Propagate the upgrate call to SCC

### PR DESCRIPTION
**Pull request by jesusbv**
## Description

Proxy the zypper migration upgrade call to SCC when the system is BYOS
* Related Issue / Ticket / Trello card: <link reference>

## How to test 

- 15 SP3 BYOS instance
- register to RMT 2.19
- add LTSS
- update
- remove LTSS
- migrate to 15 SP4
- registercloudguest -r $LTSS_REGCODE

should work

## Change Type

*Please select the correct option.*

- [ ] **Bug Fix** (a non-breaking change which fixes an issue)
- [ ] **New Feature** (a non-breaking change which adds new functionality)
- [ ] **Documentation Update** (a change which only updates documentation)

## Checklist

*Please check off each item if the requirement is met.*

- [X] I have reviewed my own code and believe that it's ready for an external review.
- [X] I have provided comments for any hard-to-understand code.
- [ ] I have documented the `MANUAL.md` file with any changes to the user experience.
- [ ] If my changes are non-trivial, I have added a changelog entry to notify users at `package/obs/rmt-server.changes`.

## Review

Please check out our [review guidelines](https://github.com/SUSE/scc-docs/blob/master/team/workflow/code_review.md) 
and get in touch with the author to get a shared understanding of the change.

### CODE REVIEW

File: engines/scc_proxy/lib/scc_proxy/engine.rb [hunk 1]
```diff
@@ -117,6 +117,17 @@ def prepare_scc_request(uri_path, product, auth, params, mode)
       scc_request
     end
 
+    def prepare_scc_update_request(uri_path, product, auth, mode)
```
Comment (jesusbv) [inline engines/scc_proxy/lib/scc_proxy/engine.rb:120]: yes, `prepare_scc_request` and `prepare_scc_update_request` can be refactored into one method

let's get make it work first, then let's make it pretty

File: engines/scc_proxy/lib/scc_proxy/engine.rb [hunk 2]
```diff
@@ -129,7 +140,7 @@ def announce_system_scc(auth, params)
     end
 
     def scc_activate_product(product, auth, params, mode)
-      uri = URI.parse(ACTIVATE_PRODUCT_URL)
+      uri = URI.parse(SYSTEM_PRODUCTS_URL)
```
Comment (rjschwei) [inline engines/scc_proxy/lib/scc_proxy/engine.rb:143]: Shouldn't activation stay the same, i.e. the code path taken when the `registercloudguest -r $SLES_REG_CODE` is being used?

Comment (jesusbv) [reply engines/scc_proxy/lib/scc_proxy/engine.rb:143]: it is the same URL, one does a post and one does a put, the URL does not mean `let's activate products` any more

**Review by rjschwei — APPROVED** [review]

File: engines/scc_proxy/lib/scc_proxy/engine.rb [hunk 3]
```diff
@@ -337,12 +362,13 @@ def scc_activate_product
...
               raise 'Incompatible extension product' unless @product.arch == base_prod.arch && @product.version == base_prod.version
 
-              params['hostname'] = @system.hostname
-              params['proxy_byos_mode'] = mode
-              params['scc_login'] = @system.login
-              params['scc_password'] = @system.password
-              params['hwinfo'] = JSON.parse(@system.system_information)
-              params['instance_data'] = @system.instance_data
+              update_params_system_info mode
+              # params['hostname'] = @system.hostname
```
Comment (digitaltom) [inline engines/scc_proxy/lib/scc_proxy/engine.rb:366]: Looks like those comments are a left over?

File: engines/scc_proxy/lib/scc_proxy/engine.rb (hunk 1 above)
Comment (digitaltom) [inline engines/scc_proxy/lib/scc_proxy/engine.rb:120]:
```suggestion
    def prepare_scc_upgrade_request(uri_path, product, auth, mode)
```

File: engines/scc_proxy/lib/scc_proxy/engine.rb [hunk 4]
```diff
@@ -254,6 +265,20 @@ def scc_check_subscription_expiration(headers, login, system_token, logger, mode
...
       SccProxy.activations_fail_state(scc_systems_activations, headers, product)
     end
+
+    def scc_upgrade(auth, product, system_login, mode, logger)
+      uri = URI.parse(SYSTEM_PRODUCTS_URL)
+      http = Net::HTTP.new(uri.host, uri.port)
+      http.use_ssl = true
+      scc_request = prepare_scc_update_request(uri.path, product, auth, mode)
+      response = http.request(scc_request)
+      unless response.code_type == Net::HTTPOK
```
Comment (digitaltom) [inline engines/scc_proxy/lib/scc_proxy/engine.rb:275]:
```suggestion
      unless response.code_type == Net::HTTPCreated
```

Comment (jesusbv) [reply engines/scc_proxy/lib/scc_proxy/engine.rb:275]: I will change the http status code, as mentioned, I followed the swagger documentation
image

**Review by digitaltom — APPROVED** [review]

**Pull request merged**